alternator/expressions.g

```diff
@@ -231,19 +250,20 @@ primitive_condition returns [parsed::primitive_condition c]:
 // common rule prefixes, and (lack of) support for operator precedence.
 // These rules could have been written more clearly using a more powerful
 // parser generator - such as Yacc.
-boolean_expression returns [parsed::condition_expression e]:
-	  b=boolean_expression_1       { $e.append(std::move($b.e), '|'); }
-	  (OR b=boolean_expression_1   { $e.append(std::move($b.e), '|'); } )*
+// See comment above why the "depth" counter was needed here.
+boolean_expression[int depth] returns [parsed::condition_expression e]:
+	  b=boolean_expression_1[depth]       { $e.append(std::move($b.e), '|'); }
+	  (OR b=boolean_expression_1[depth]   { $e.append(std::move($b.e), '|'); } )*
 	;
-boolean_expression_1 returns [parsed::condition_expression e]:
-	  b=boolean_expression_2       { $e.append(std::move($b.e), '&'); }
-	  (AND b=boolean_expression_2  { $e.append(std::move($b.e), '&'); } )*
+boolean_expression_1[int depth] returns [parsed::condition_expression e]:
+	  b=boolean_expression_2[depth]       { $e.append(std::move($b.e), '&'); }
+	  (AND b=boolean_expression_2[depth]  { $e.append(std::move($b.e), '&'); } )*
 	;
-boolean_expression_2 returns [parsed::condition_expression e]:
+boolean_expression_2[int depth] returns [parsed::condition_expression e]:
 	  p=primitive_condition        { $e.set_primitive(std::move($p.c)); }
-	| NOT b=boolean_expression_2   { $e = std::move($b.e); $e.apply_not(); }
-	| '(' b=boolean_expression ')' { $e = std::move($b.e); }
+	| {depth<MAX_DEPTH}? NOT b=boolean_expression_2[depth+1]   { $e = std::move($b.e); $e.apply_not(); }
+	| {depth<MAX_DEPTH}? '(' b=boolean_expression[depth+1] ')' { $e = std::move($b.e); }
```

Comment:
just an idea, if I understand correctly how this works you can add something like (my syntax may not be fully correct here):

```
| {depth>=MAX_DEPTH} { throw std::expressions_syntax_error("max depth of 400 reached"); }
```

to improve error message

Reply:
Nice idea. I'll try this. Of course it adds even more boilerplate garbage to the parser specification, which is sad, but it's already ugly enough with all those "depth", so two more lines won't hurt anyone. I'll try.

Reply:
I have to admit, that I tried to do this and failed :-( Apparently, the "?" doesn't just turn off the rule if it's false - it throws an exception of the FAILED_PREDICATE_EXCEPTION immediately. There is supposed to be a "?=>" syntax instead of "?" that doesn't do this, but I couldn't get it to work either. This Antlr is a big steaming heap of @#$!...

I'll try to see if I can get the code that prints a syntax error to print a FAILED_PREDICATE_EXCEPTION (if it even works) in a more useful way.

Reply:
The result came out very nice and clean, I'll send a new version in a few minutes.